rfd/0067-database-access-aws-redis

```diff
@@ -0,0 +1,221 @@
+
+Here is a sample config for specifying the secret store in database service.
+```
+db_service:
+  ...
+
+  secert_store:
+    type: "aws-secrets-manager"
+    key_prefix: "teleport/"
+```
+
+And here is a sample interface for managing secrets using the secret store.
+```
+// SecretValue is the secret value.
+type SecretValue struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret.
+    Value string
+    // Version is the version of the secret value.
+    Version string
+    // Created is the creation time of this version. Can be used to calculate
+    // if a secret value is expired.
+    Created time.Time
+}
+
+// CreateRequest defines the request to create a new secret.
+type CreateRequest struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+}
+
+// PutValueRequest defines the request to create a new secret version.
+type PutValueRequest struct {
+    // CurrentValue is the current secret value. Key in CurrentValue is used to
+    // identify the secret. Other attributes can be used for test-and-set
+    // operations.
+    CurrentValue *SecretValue
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+}
+
+// Secrets defines an interface for managing secrets. A secret consists of a
+// key path and a list of versions which hold copies of current or past secret
+// values.
+type Secrets interface {
+    // Create creates a new secret.
+    Create(req CreateRequest) (*SecretValue, error)
+
+    // Delete deletes the secret for provided path.
+    Delete(key string) error
+
+    // PutValue creates a new secret version for the secret. Implementations
+    // must guarantee that only the first of many calls to PutValue succeeds,
+    // when req.CurrentValue are the same.
+    PutValue(req PutValueRequest) (*SecretValue, error)
+
+    // GetValue returns the secret value for provided version. Besides version
+    // string returned from PutValue, two specials versions "CURRENT" and
+    // "PREVIOUS" can also be used to retrieve the current and previous
+    // versions respectively. If version is empty, "CURRENT" is used.
+    GetValue(key string, version string) (*SecretValue, error)
+}
+```
+It is up to the caller to call `PutValue` periodically to rotate the secret.
+
+#### Secret store - Random password generation
+[sethvargo/go-password](https://github.com/sethvargo/go-password) can be used to generate random passwords. <sup>[3]<sup>
```

Comment:
FWIW we have utils.CryptoRandomPassword that generates a cryptographically secure random string. Should be ok to use it and not introduce extra dependencies.

Reply:
Where is this function? I tried looking for it, but couldn't find it.

Reply:
Is this the one?
https://github.com/gravitational/teleport/blob/b24b561560707161d4b75dbecafeb83c5ee3c7d4/lib/utils/rand.go#L28-L33

This one generates hex ([0-9] and [a-f]).

Reply:
I've removed this section if `CryptoRandomHex` is good to use.

Reply:
@reedloden I have updated the security section of this RFD. For Redis users, the max password length is 128 and I am planning to use the above function for password generation. Please review if you have any comments on that. Thanks!!

Reply:
Even with a smaller keyspace from only hex characters, a 128 character password has plenty of entropy, so should be fine. Plus, I personally try to avoid special characters in database-like passwords, as things have a tendency to get interpreted incorrectly (especially when you don't control both input and output). So, sounds good to me.